File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
private:
    ListNode* mergeTwoList(ListNode* list1, ListNode* list2) {
        ListNode* dummy = new ListNode();
        ListNode* node = dummy;
        while (list1 && list2) {
            if (list1->val <= list2->val) {
                node->next = list1;
                list1 = list1->next;
                node = node->next;
            } else {
                node->next = list2;
                list2 = list2->next;
                node = node->next;
            }
        }
        if (list1) node->next = list1;
        else node->next = list2;
        return dummy->next;
    }
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) return nullptr;
        while (lists.size() > 1) {
            // Devide and Conquer
            vector<ListNode*> mergedLists;
            for (int i = 0; i < lists.size(); i+=2) {
                if (i + 1 >= lists.size()) {
                    mergedLists.push_back(lists[i]);
                    break;
                }
                mergedLists.push_back(
                    mergeTwoList(lists[i], lists[i + 1]));
            }
            lists = mergedLists;
        }
        return lists.back();
    }
};

File: Data Structures & Algorithms/add-two-numbers/submission-3.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} l1
     * @param {ListNode} l2
     * @return {ListNode}
     */
    addTwoNumbers(l1, l2) {
        const dummy = new ListNode();
        let node = dummy, carry = 0;
        while (l1 || l2 || carry) {
            let l1_val = l1 ? l1.val : 0;
            let l2_val = l2 ? l2.val : 0;
            let summary = l1_val + l2_val + carry;
            node.next = new ListNode(summary % 10);
            node = node.next;
            carry = Math.floor(summary / 10);
            l1 = l1 ? l1.next : l1;
            l2 = l2 ? l2.next : l2;
        }
        return dummy.next;
    }
}
